File: typescript/src/monitor.ts
import { v4 as uuidv4 } from "uuid";
import { z } from "zod";

export const MonitorCreateSchema = z.object({
  name: z.string(),
  url: z.string().url(),
  interval_seconds: z.number().int().min(5),
  timeout_ms: z.number().int().min(100),
  failure_threshold: z.number().int().min(1),
});
export type MonitorCreateRequest = z.infer<typeof MonitorCreateSchema>;

export type MonitorStatus = "UNKNOWN" | "UP" | "DOWN";

export interface Monitor {
  id: string;
  name: string;
  url: string;
  interval_seconds: number;
  timeout_ms: number;
  failure_threshold: number;
  status: MonitorStatus;
  consecutive_failures: number;
  created_at: string;
  updated_at: string;
}

export function createMonitor(req: MonitorCreateRequest): Monitor {
  const now = new Date().toISOString();
  return {
    id: uuidv4(),
    name: req.name,
    url: req.url,
    interval_seconds: req.interval_seconds,
    timeout_ms: req.timeout_ms,
    failure_threshold: req.failure_threshold,
    status: "UNKNOWN",
    consecutive_failures: 0,
    created_at: now,
    updated_at: now,
  };
}

// TODO: recordResult(monitor, success) — status transition logic per Section 4

File: go/monitor.go
package main

import (
	"time"

	"github.com/google/uuid"
)

type Monitor struct {
	ID                  string    `json:"id"`
	Name                string    `json:"name"`
	URL                 string    `json:"url"`
	IntervalSeconds     int       `json:"interval_seconds"`
	TimeoutMs           int       `json:"timeout_ms"`
	FailureThreshold    int       `json:"failure_threshold"`
	Status              string    `json:"status"` // UNKNOWN | UP | DOWN
	ConsecutiveFailures int       `json:"consecutive_failures"`
	CreatedAt           time.Time `json:"created_at"`
	UpdatedAt           time.Time `json:"updated_at"`
}

func NewMonitor(name, url string, interval, timeoutMs, threshold int) *Monitor {
	now := time.Now().UTC()
	return &Monitor{
		ID:               uuid.NewString(),
		Name:             name,
		URL:              url,
		IntervalSeconds:  interval,
		TimeoutMs:        timeoutMs,
		FailureThreshold: threshold,
		Status:           "UNKNOWN",
		CreatedAt:        now,
		UpdatedAt:        now,
	}
}

// TODO: Validate() (*errors.New style, wrapped) — interval >= 5, timeout >= 100, etc.
// TODO: RecordResult(success bool) — status transition logic per Section 4
